redhat-appstudio/service-provider-integration-operator
# add tokenUpload controller

### What does this PR do?
It enables upload of Personal Access Token using very short living K8s Secret.
Controller recognizes the Secret by label.spi.appstudio.redhat.com/token: "" label, gets the PAT and the Name of the SPIAccessToken associated with it, then deletes the Secret (for better security reason) and uploads the Token (which in turn updates the Status of associated SPIAccess/Token/TokenBinding).

### What issues does this PR fix or reference?

https://issues.redhat.com/browse/SVPI-273

### How to test this PR?
No automated tests provided for the time (TODO)
To test it manually:
- Run operator with  ENABLETOKENUPLOAD=true. For example off-cluster it may look like:  
make run_as_current_user VAULTHOST=https://vault.192.168.64.6.nip.io VAULTAPPROLEROLEIDFILEPATH=.secretdata/role_id VAULTAPPROLESECRETIDFILEPATH=.secretdata/secret_id VAULTINSECURETLS=true CONFIGFILE=.secretdata/config.yaml ENABLETOKENUPLOAD=true
- Create spiaccesstokenbinding object with SP of your choice. For example: spec.repoUrl: https://github.com for GitHub 
- Check this spiaccesstokenbinding, like: kubectl get spiaccesstokenbinding <name> -o yaml . It's status.phase should be AwaitingTokenData
- Take Linked Token Name from status.linkedAccessTokenName. 
- Get Service Provider's Personal Access Token. For example from GitHub->Settings->Developer Settings->Personal Access Tokens
- Create the Secret like:
```
apiVersion: v1
kind: Secret
metadata:
  name: test-secret
  labels:
    spi.appstudio.redhat.com/upload-secret: token
    spi.appstudio.redhat.com/token-name: <spi-access-token-name>
type: Opaque
stringData:
  tokenData: <serviceprovider-pat>
```
- Check this spiaccesstokenbinding, like: kubectl get spiaccesstokenbinding <name> -o yaml . It's status.Phase should be Injected
